src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -0,0 +1,93 @@
+/*******************************************************************\
+
+ Module: Generate Java Generic Type - Instantiate a generic class with
+         concrete type information.
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+#include "generate_java_generic_type.h"
+#include <util/namespace.h>
+#include <java_bytecode/java_types.h>
+#include <java_bytecode/java_utils.h>
+
+#include <iostream>
+#include <sstream>
+
+generate_java_generic_typet::generate_java_generic_typet(
+  symbol_tablet &symbol_table,
+  message_handlert &message_handler):
+    symbol_table(symbol_table),
+    message_handler(message_handler)
+{}
+
+symbolt generate_java_generic_typet::operator()(
+  const java_generic_typet &existing_generic_type)
+{
+  PRECONDITION(existing_generic_type.id()==ID_pointer);
+  namespacet ns(symbol_table);
+
+  const typet &pointer_subtype=ns.follow(existing_generic_type.subtype());
+
+  INVARIANT(
+    pointer_subtype.id()==ID_struct, "Only pointers to classes in java");
+
+  const java_class_typet &java_class=to_java_class_type(pointer_subtype);
+  java_class_typet replacement_type=java_class;
```

Comment:
I seem to have used two different ways of copying the class - first we copy the existing one and modify the components, then we generate a new stub and use the components. I think you should do only one. I think the `generate_class_stub` is probably better, so you should just build a `componentst` here rather than a whole type.